Remove package extensions in Pkg functions

fredrikekre opened the pull request:
This implements #20630 and allows for eg.

```julia
Pkg.add("SpecialFunctions.jl")
Pkg.rm("SpecialFunctions.jl")
...
```
by stripping of the terminating `.jl` extension.

TODO:
- [x] add tests (ideas?)
- [x] documentation (where should this be documented?)

fix #20630

fredrikekre committed:
remove package extension in Pkg functions

ararslan commented:
Thanks for working on this! As a test, you could add/remove the Example package both with and without the .jl suffix.

tkelman commented on test/pkg.jl:
```diff
@@ -496,6 +496,53 @@ temp_pkg_dir() do
+        Pkg.status("Example.jl", iob)
+        str = chomp(String(take!(iob)))
+        @test endswith(str, string(Pkg.installed("Example.jl")))
+        @test isempty(Pkg.dependents("Example.jl"))
```
this seems a little fragile, anyone could break this by adding a package that depends on Example in the future (which is kinda silly, but allowed)

fredrikekre replied:
I just copied from the usual test suite: https://github.com/JuliaLang/julia/blob/1297ede6cf40e0ad7e893fc4a2916a20a3df78ce/test/pkg.jl#L127

fredrikekre replied:
But yeah, both tests seem fragile.

StefanKarpinski replied:
Seems pretty hypothetical. If anyone does that, we can fix the tests then.

tkelman replied:
That would immediately break the tests of all past versions of julia though, which isn't great. Would be better to move this to a section of the tests that's running against a known past commit of metadata, rather than the latest.

fredrikekre replied:
Feel free to push a commit to this branch fixing this and the test linked in my comment above or handle it in a separate PR. I am not sure how I should go about it.

yuyichao commented on base/pkg/pkg.jl:
```diff
@@ -37,6 +37,10 @@ const cd = Dir.cd
 
+# remove extension .jl
+const PKGEXT = ".jl"
+splitjl(pkg::AbstractString) = endswith(pkg, PKGEXT) ? splitext(pkg)[1] : pkg
```
Not that it really matters but `pkg[1:(end - 3)]` would be much faster...

ararslan commented on doc/src/manual/packages.md:
```diff
@@ -171,6 +171,14 @@ git config --global url."https://".insteadOf git://
 However, this change will be system-wide and thus the use of [`Pkg.setprotocol!()`](@ref) is preferable.
 
+!!! note
+    The package manager functions accept package names with the `.jl` extension:
```
I'd probably word this to say that it's possible but is stripped out internally so as not to encourage this style of use. Maybe "The package manager functions also accept the `.jl` suffix on package names, though the suffix is stripped out internally."

fredrikekre commented:
Failures seems unrelated? Feel free to restart.

KristofferC commented:
Restarded travis, log at: https://gist.github.com/KristofferC/162cbe28353847ad4197e5e9c2337454

tkelman committed:
Run Pkg.dependents("Example") tests at a fixed METADATA commit

to avoid fragility in the case that any future package
happens to depend on Example.jl

fredrikekre commented:
Bump. (Too late for this to be merged?)
